### ISSUE TITLE
Feature: job.inputs

### ISSUE THREAD

**Issue by jonathanhefner** [evanphx/benchmark-ips#70]
TL;DR: It would be cool to set an array of inputs as part of benchmark job config, and have every input yielded to each `report` block.

##### Motivating Example
Let's say you're benchmarking different regular expressions.  You'll preferably try each variant on not just one, but a set of input strings:
```ruby
STRINGS = ["fool", "founded", "funded", "foolish"]
REGEX1 = /fo+/
REGEX2 = /foo*/

Benchmark.ips do |job|
  job.report("regex1"){ STRINGS.each{|s| s =~ REGEX1 } }
  job.report("regex2"){ STRINGS.each{|s| s =~ REGEX2 } }
  job.compare!
end
```

This works, but it's a tad messy.  You could DRY it up by defining `regex1` and `regex2` methods, but that adds some marginal amount of performance overhead.  On the topic of overhead, though, those `STRINGS.each` calls may be a problem.  You could eliminate the `each` loop using careful string concatenation:
```ruby
STRINGS = ["fool", "founded", "funded", "foolish"]
REGEX1 = /fo+/
REGEX2 = /foo*/

R1 = STRINGS.each_index.map{|i| "STRINGS[#{i}] =~ REGEX1" }.join(";")
R2 = STRINGS.each_index.map{|i| "STRINGS[#{i}] =~ REGEX2" }.join(";")

Benchmark.ips do |job|
  job.report("regex1", R1)
  job.report("regex2", R2)
  job.compare!
end
```

But that's even uglier, and more error-prone, and if your expressions boil down to e.g. `==` comparison, the interpreter may throw warnings at you like "possibly useless use of == in void context."

##### Proposed Solution

It would be nice if you could do something like the following:
```ruby
REGEX1 = /fo+/
REGEX2 = /foo*/

Benchmark.ips do |job|
  job.inputs = ["fool", "founded", "funded", "foolish"]
  job.report("regex1"){|s| s =~ REGEX1 }
  job.report("regex2"){|s| s =~ REGEX2 }
  job.compare!
end
```

Preserving the semantics of the first example, while not counting the loop overhead in the run time.

##### Other Benefits

Tasks which destructively modify their inputs can be difficult to accurately benchmark, because cloning their inputs between iterations can skew run times.  (See JuanitoFatas/fast-ruby#101 for an example of this.)  Given a reference to the inputs, the job could perform the necessary clone, also excluding it from the run times.  For efficiency, this behavior could be disabled by default, and enabled via e.g. `job.destructive? = true`.

**Comment by evanphx** [comment]
I don't see a reason to add this, honestly. You've got a good use case but you're best wrapping it up with your own code rather than trying to bake a generic version into `benchmark-ips`.

**Comment by jonathanhefner** [comment]
In case anyone else finds it useful, I've packaged this feature into a [benchmark-inputs](https://github.com/jonathanhefner/benchmark-inputs) gem.  The syntax is similar to the above proposal:

```ruby
Benchmark.inputs(["fool", "founded", "funded", "foolish"]) do |job|
  job.report("regex1"){|s| s =~ /fo+/ }
  job.report("regex2"){|s| s =~ /foo*/ }
  job.compare!
end
```

Benchmarking destructive operations also works:

```ruby
Benchmark.inputs(["fool", "founded", "funded", "foolish"]) do |job|
  job.dup_inputs = true  # <--- for destructive operations
  job.report("regex1"){|s| s.gsub!(/fo+/, '') }
  job.report("regex2"){|s| s.gsub!(/foo*/, '') }
  job.compare!
end
```

**Comment by kbrock** [comment]
thanks @jonathanhefner
